src/main/java/org/mastodon/mamut/tomancak/lineage_registration/EstimateTransformation.java

```diff
@@ -0,0 +1,59 @@
+package org.mastodon.mamut.tomancak.lineage_registration;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import mpicbg.models.AbstractAffineModel3D;
+import mpicbg.models.IllDefinedDataPointsException;
+import mpicbg.models.NotEnoughDataPointsException;
+import mpicbg.models.Point;
+import mpicbg.models.PointMatch;
+import mpicbg.models.SimilarityModel3D;
+import net.imglib2.RealPoint;
+import net.imglib2.realtransform.AffineTransform3D;
+import org.mastodon.collection.RefRefMap;
+import org.mastodon.mamut.model.Spot;
+
+public class EstimateTransformation
+{
+
+	/**
+	 * Return a affine transform, that is composed of scaling, rotation and translation operation.
+	 * The transformation is optimized to minimize the distances of the transformed "key" spots
+	 * to the "value" spots.
+	 */
+	public static AffineTransform3D estimateScaleRotationAndTranslation( RefRefMap< Spot, Spot > pairs )
+	{
+		List< RealPoint > pointsA = new ArrayList<>();
+		List< RealPoint > pointsB = new ArrayList<>();
+		for( Spot rootA : pairs.keySet() ) {
+			Spot rootB = pairs.get( rootA );
+			pointsA.add( new RealPoint( rootA ) );
+			pointsB.add( new RealPoint( rootB ) );
+		}
+		return estimateScaleRotationAndTranslation( pointsA, pointsB );
+	}
+	
+	static AffineTransform3D estimateScaleRotationAndTranslation( List< RealPoint > a, List< RealPoint > b )
+	{
+		AbstractAffineModel3D model = new SimilarityModel3D();
+		assert a.size() == b.size();
+		List< PointMatch > matches = new ArrayList<>(a.size());
+		for ( int i = 0; i < a.size(); i++ )
+		{
+			matches.add(new PointMatch( new Point( a.get( i ).positionAsDoubleArray() ), new Point( b.get( i ).positionAsDoubleArray() ), 1 ));
```

Comment:
I reduced the amount of memory allocation that happens in the EstimateTransform class.
But bringing it down to a level that fits what we normally do in Mastodon would require a rewrite the SimilarityModel3D class. The class is part of another library (the maven artifact mpicbg:mpicbg:1.4.2). I will do the rewrite if performance problems actually show up.

Reply:
thank you

yes, since `SimilarityModel3D` is outside, we just have to take it as it is... (and fix it only when it becomes a real problem). But the twist in our code was in our hands, so we could easily do something about it... I think you did great when you just switched it for a better code (performance-wise).